[fmiweather] FMI Weather Binding initial contribution

ssalonen opened the pull request:
- [x] ~~currently I use java system default timezone for formatting ZonedDateTime for display. Should I use `org.eclipse.smarthome.core.i18n.TimeZoneProvider` instead? I can find both ways implemented in other addons.~~ Decided to keep using system default timezone, see below discussion.
- [x] bump snapshot version to match `2.5.x`
- [ ] merge docs logo https://github.com/openhab/openhab-docs/pull/929

**To try out the binding, one should use 2.5.0-SNAPSHOT due to UoM changes in core.**

This PR replaces #5226. Should now be follow the new build system.

Signed-off-by: Sami Salonen <ssalonen@gmail.com>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

ssalonen commented:
There's still something wrong... Any tips @Hilbrand?

I tried my best to follow the tutorial you linked in #5226 :(
``` 
Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-fmiweather; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-fmiweather)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-fmiweather/2.5.0.SNAPSHOT: missing requirement [openhab-binding-fmiweather/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.fmiweather; type=osgi.bundle; version="[2.5.0.201911021648,2.5.0.201911021648]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.fmiweather/2.5.0.201911021648: missing requirement [org.openhab.binding.fmiweather/2.5.0.201911021648] osgi.wiring.package; filter:="(osgi.wiring.package=org.openhab.io.net.http)"]]
Repositories: {
	file:/home/travis/build/openhab/openhab2-addons/bundles/org.openhab.binding.fmiweather/target/feature/feature.xml
	mvn:org.apache.karaf.features/framework/4.2.6/xml/features
	mvn:org.apache.karaf.features/standard/4.2.6/xml/features
	mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/2.5.0-SNAPSHOT/xml/features
	mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-tp/2.5.0-SNAPSHOT/xml/features
	mvn:org.ops4j.pax.web/pax-web-features/7.2.10/xml/features
}
```

Hilbrand commented:
I would guess you need to add to the feature.xml:
```
        <feature>openhab-transport-http</feature>
```

ssalonen commented:
> I would guess you need to add to the feature.xml:
> 
> ```
>         <feature>openhab-transport-http</feature>
> ```

Thanks @Hilbrand! Unfortunately still get the same error (see TravisBuddy above)

ssalonen commented:
@Hilbrand reverted the change with openhab-transport-http since it had no effect.

I looked more closer to other bindings using HttpUtil and noticed a detail I missed before: the other bindings `import org.eclipse.smarthome.io.net.http.HttpUtil` while I had `import org.openhab.io.net.http.HttpUtil`. This seemed to make a difference, and now I can locally build the bundle.

I expect jenkins to succeed this time as well.

ssalonen commented:
Allright, managed to get the build working and binding seems to work just fine in the demo.

Only one open question 

> currently I use java system default timezone for formatting ZonedDateTime for display. Should I use org.eclipse.smarthome.core.i18n.TimeZoneProvider instead? I can find both ways implemented in other addons.

kaikreuzer commented:
> currently I use java system default timezone for formatting ZonedDateTime for display. Should I use org.eclipse.smarthome.core.i18n.TimeZoneProvider instead? I can find both ways implemented in other addons.

Good question, where I am myself not sure about the answer.
@openhab/core-maintainers Can any of you answer? I would expect that the binding simply provides the datetime, specifying which time zone it refers to, while the UIs show datetime values always in the time zone the user has specified - but I am not certain, if this is how it currently works?

cweitkamp commented:
This is a really good question. Currently the `TimeZoneProvider` is used for persistence and `ChartProvider` implementations (https://github.com/openhab/openhab-core/search?q=TimeZoneProvider&unscoped_q=TimeZoneProvider). For UIs the `DateTimeType` values will be converted to the locale timezone - which IIRC is the system default timezone (see [ItemUIRegistryImpl](https://github.com/openhab/openhab-core/blob/07d10dba3d73349ce2f302a6393e255ee4de788d/bundles/org.openhab.core.ui/src/main/java/org/eclipse/smarthome/ui/internal/items/ItemUIRegistryImpl.java#L414-L420)).

In my binding implementations I am used to convert the datetime values delivered by the service provider(s) to the system default timezone.

wborn commented:
It looks like the configured time zone really is the system time and is used as a more convenient way to configure the time zone instead of adding `-Duser.timezone=` as JVM startup parameter. 

We could then also use it to update the system default time zone using [TimeZone.setDefault](https://docs.oracle.com/javase/8/docs/api/java/util/TimeZone.html#setDefault-java.util.TimeZone-). Then there would be no need to use the provider for it at all and everything will be in sync.

Updating the time zone with `TimeZone.setDefault` may also prevent issues when using certain frameworks or libraries which also use this value.

ssalonen commented:
OK, thanks for the answers. I will then keep the current implementation which will emit `DateTimeType(ZonedDateTime)` where ZoneDateTime is expressed in system default (`ZoneId.systemDefault()` -- [which delegates to `TimeZone.getDefault()`](https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html#systemDefault--)). No need to use `TimeZoneProvider`.

Or in code
```
updateState(channelUID, new DateTimeType(ZonedDateTime.ofInstant(Instant.ofEpochSecond(epochSecond), UTC)
                    .withZoneSameInstant(ZoneId.systemDefault())));
```

(I get `epochSecond` in UTC from the weather API)

>  I would expect that the binding simply provides the datetime, specifying which time zone it refers to, while the UIs show datetime values always in the time zone the user has specified

This would be ideal I think but I think not matching my observations before. It's been months already though, and I am not sure if it was PaperUI or BasicUI I was checking. I was initially just emitting ZonedDateTime that are expressed in UTC (-> UI showed UTC times), later on changed them to use system default timezone (using [ZoneDateTime.withZoneSameInstant](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#withZoneSameInstant-java.time.ZoneId-) to do the timezone conversion from UTC).

ssalonen commented:
For the record, the code is complete from my side, and ready for further reviews. There were already user testing reports in the previous PR (#5226).

I've now addressed the build system issues (bnd conversion) and issues identified by SAT. Everything seems to be working as expected.


Let me know if there's anything I can do to ease the review work.

ssalonen commented:
Branch changed to `2.5.x` as adviced in #6702

bjoernbrings commented:
Have exactly the same problem travis has
```
[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.2.7:verify (karaf-feature-verification) on project org.openhab.binding.fmiweather: Unable to load features descriptors: Error:
[ERROR]         Error downloading mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/2.5.4-SNAPSHOT/xml/features
```

Tried to follow up on the reason but could not find it. I guess it's some dependency.

cpmeister commented on bundles/org.openhab.binding.fmiweather/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<features name="org.openhab.binding.fmiweather-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+    <repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${project.version}/xml/features</repository>
```
This should fix your dependency problem.
```suggestion
    <repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
```

bjoernbrings replied:
That works for my setup. Thanks @cpumeister for the debugging

ssalonen replied:
Much appreciated!

bjoernbrings commented:
Did a first test and recognize, that I tested this binding before as I'm struggeling with the same finding as before.

Compiles and works.

Currently the last forecast with data is 45h, everything beyond is NaN. Wouldn't it be better to ignore the ranges where there is the possibility of having no data?

I'm not through with the code. But one (minor) thing that I saw quite often were static functions that were not declared static. Is that by intention?

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,213 @@
+            // IllegalStateException: Unexpected (possibly bug) issue with response
+            response = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    String.format("{}: {}", e.getClass().getName(), e.getMessage()));
```
I'm getting "Status: OFFLINE - COMMUNICATION_ERROR {}: {}" isn't it String.format("%1$s: %2$s", ...)?

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,213 @@
+            }
+        } catch (IllegalStateException e) {
+            // IllegalStateException: Unexpected (possibly bug) issue with response
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
See comment above

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,213 @@
+        }
+    }
+
+    private int getTimeIndex(int hours) {
```
One out of several functions that I would consider to be static

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,382 @@
+				<label>FMISID of the weather station</label>
+				<description><![CDATA[Station ID (FMISID) of the weather observation station
+				<br />
+				<br />See https://en.ilmatieteenlaitos.fi/observation-stations for a list of observation stations]]></description>
```
Worth to mention that only ids of 'weather' stations do make sense here?

ssalonen replied:
Clarified instructions in d917bdc. In principle one can pick e.g. precipitation observation station and only that data would show up.

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,382 @@
+	</thing-type>
+
+
+	<!-- Groups -->
```
How did you create the selection of channels? As I looked into the referenced [webpage(s)](http://opendata.fmi.fi/meta?observableProperty=forecast) there are many more channels.

ssalonen replied:
It's a good point, I think I picked the most "common" / "relevant" channels but perhaps I could pick all.

In the document they have an example query similar to below, not restricting any parameters in the call:
http://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=fmi::forecast::hirlam::surface::point::multipointcoverage&place=jaala&latlon=60.1,19.9&

Resulting in these parameters (bold ones included in the binding):

1. GeopHeight [not related to forecasts really]
2. **Temperature**
3. **Pressure**
4. **Humidity**
5. **WindDirection**
6. **WindSpeedMS**
7. WindUMS  [redundant if one has direction and m/s speed]
8. WindVMS  [redundant if one has direction and m/s speed]
9. MaximumWind [maximum taken from 6 short-term-speed-averages-of-10min?]
10. **WindGust**
11. DewPoint [would be useful?]
12. **TotalCloudCover**
13. **WeatherSymbol3**
14. LowCloudCover [does user care about this detailed cloud information when TotalCloudCover is available?]
15. MediumCloudCover [does user care about this detailed cloud information when TotalCloudCover is available?]
16. HighCloudCover [does user care about this detailed cloud information when TotalCloudCover is available?]
17. **Precipitation1h**
18. PrecipitationAmount [what is the meaning of this?]
19. RadiationGlobalAccumulation
20. RadiationLWAccumulation
21. RadiationNetSurfaceLWAccumulation
22. RadiationNetSurfaceSWAccumulation
23. RadiationDiffuseAccumulation
24. LandSeaMask [not related to forecasts really]

The [API](http://opendata.fmi.fi/meta?observableProperty=forecast) you linked gives even more parameters such as 
```xml
<component>
  <ObservableProperty gml:id="cloud_indicator">
      <label>Lightning type indicator</label>
      <basePhenomenon>Lightning</basePhenomenon>
      <uom uom="Index"/>
  </ObservableProperty>
</component>
```

But it is not queriable at least with `storedquery_id=fmi::forecast::hirlam::surface::point::multipointcoverage`:
http://opendata.fmi.fi/wfs?service=WFS&version=2.0.0&request=getFeature&storedquery_id=fmi::forecast::hirlam::surface::point::multipointcoverage&place=jaala&latlon=60.1,19.9&parameters=cloud_indicator

```xml
<Exception exceptionCode="OperationParsingFailed">

   <ExceptionText>Unknown parameter &apos;cloud_indicator&apos;!</ExceptionText>
   <ExceptionText>URI: /wfs?latlon=60.1%2C19.9&amp;parameters=cloud_indicator&amp;place=jaala&amp;request=getFeature&amp;service=WFS&amp;storedquery_id=fmi%3A%3Aforecast%3A%3Ahirlam%3A%3Asurface%3A%3Apoint%3A%3Amultipointcoverage&amp;version=2.0.0</ExceptionText>
    
</Exception>
```

So I'm not sure what is the full set of available parameters but certainly we could add more.

Do you see anything relevant?


P.S. as side note I have been pondering of adding additional FMI data sources such as air quality, road condition warnings etc. but perhaps that is for a later date.

bjoernbrings replied:
Honestly I was coming across that as I was (incorrectly) providing some non-'weather' stations and was asking myself whether they could provide some usefull observations for Finnish users.
Regarding the forecast I would agree that it felt to me like you selected the most important ones but was not sure whether there were probably forecasts on some topics that are not provided by other weather bindings.

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,382 @@
+	<channel-type id="present-weather-channel">
+		<item-type>Number</item-type>
+		<label>Prevailing weather</label>
+		<state readOnly="true" min="0" max="99" pattern="%s">
```
I guess a short description (like the comment in the xml) will help here a lot :)

ssalonen replied:
Fixed in 4d2aadb

bjoernbrings commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,382 @@
+	<channel-type id="weather-id-channel">
+		<item-type>Number</item-type>
+		<label>Prevailing weather id</label>
+		<category>Pressure</category>
```
Does the category match? A description might be helpful here as well.

ssalonen replied:
Fixed in 4d2aadb, please review if the description is OK

ssalonen commented:
Thank you very much @bjoernbrings for the review! Resolved most of the comments.

Regarding forecast horizon -- I will monitor how often we will have NaNs in the end of the horizon and what would be more reasonable interval. It's a fair point that it might not be very useful to have 54 h horizon if most of the time the last data point is unavailable. 

For the record, our previous discussion on this topic was here: https://github.com/openhab/openhab-addons/pull/5226#issuecomment-502380028

bjoernbrings commented:
Monitored that for some days with some spot checks. At least for my location in Germany the forecast is mostly -NaN for 48h and above (which is fine, but probably something like experimental / to hide for normal users in PaperUI / advise quite prominent in the Readme). What do you think?

ssalonen commented:
@bjoernbrings 

Here's the result of by sampling the FMI forecast every 15 minutes for a bit over 4 days:
![image](https://user-images.githubusercontent.com/315866/80119802-d8f6e500-8592-11ea-85d2-4bbf20dbef54.png)

You can see that we always have forecasts available for "now" to "now + 44 h". At most we have forecasts up till "now + 50 h".

In more statistically thinking, the average/median forecast is valid up to "now + 47 h", meaning that forecasts of 48 h and beyond are halft of the time unavailable.

You can clearly see the forecast updates during the day, at 04:00, 10:00, 16:00, 22:00 UTC. As a side note, this is actually not matching what FMI is saying on their website ([Finnish](https://ilmatieteenlaitos.fi/avoin-data-saaennustedata-hirlam)) of updating it at 00, 06, 12 and 18 UTC -- seems to be they are updating at UTC+2 ;)

Furthermore, it is certainly clear that the whole 54 h (claimed by FMI) forecast does not seem to be available via the API endpoint used by the binding -- at most we get 50 h forecasts.


We certainly should remove channels beyond 50 h. Should mark forecasts from 45h and beyond as "advanced" -- hidden in PaperUI by default. There's notice in README already but I can clarify it further. I think this would work, if it is what you meant?

bjoernbrings commented:
Thanks for the nice statistics 😃

ssalonen commented:
> Does the API actually provide data for each hour (could find that hourly forecasts on FMI's homepage)?

Yes in my tests data came in every hour, only last data points were missing.

The API gives the data with the requested time resolution. Not sure what is the "native" resolution or whether results are interpolated.

In the phone app they provide hourly forecasts for the near future, and after that it is every 3h up to ~10 days in the future.

bjoernbrings commented:
So you could include forecasts for every hour (not only every 3h)? Why don't we do that?
Could be usefull to produce some nice graphs.

ssalonen commented:
I did not even remember that it was every 3h. Good suggestion!

bjoernbrings commented:
Thanks for the changes. One little thing: When I add the thing in PaperUI it still has a 3h-clustering (with all other hours-forecast going to other), so following clusters are displayed: "Forecast for the current time", "3 Hours Forecast", ..., "48 Hours Forecast", "Others" (with 1h, 2h, 4h, 5h, 7h, ... results)
Couldn't find the reason in the code.

ssalonen commented:
you have to recreate the thing for openhab to rebuild channel group descriptions I think.

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,193 @@
+                    updateEpochSecondStateIfLinked(channelUID, data.timestampsEpochSecs[lastValidIndex]);
+                } else {
+                    String field = getDataField(channelUID);
+                    @Nullable
```
See above

ssalonen replied:
Resolved: 85c70aa38087bdd184b879bcdfface2a5adf434b

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    @SuppressWarnings({ "unused", "null" })
+    private static @Nullable String getDataField(ChannelUID channelUID) {
```
See above

ssalonen replied:
I don't understand the question. No -- it's not necessary to use static. But I consider it good practice since it is not using instance data.

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,221 @@
+    protected static final String PROP_REGION = "region";
+    private static final long REFRESH_THROTTLE_MILLIS = 10_000;
+
+    protected static final int TIMEOUT_MILLIS = 30_000;
```
Can you make this way shorter to not block the scheduler thread for so long time, when the HTTP request times out?

ssalonen replied:
Reduced to 10 s: 44407132cec7a941561e50135cb85930afa90440

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Client.java:
```diff
@@ -0,0 +1,434 @@
+            locations[i] = new Location(names[i], id, latlon[0], latlon[1]);
+        }
+
+        logger.trace("names ({}): {}", names.length, names);
```
Are these trace messages useful for any user or could you use a debugger instead?

ssalonen replied:
I would like to leave these to troubleshoot user issues (results depend on time, cannot necessarily be easily reproduced by a developer afterwards)

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMIResponse.java:
```diff
@@ -0,0 +1,107 @@
+        public FMIResponse build() {
+            Map<Location, Map<String, Data>> out = new HashMap<>(timestampsByLocationByParameter.size());
+
+            timestampsByLocationByParameter.entrySet().stream().forEach(entry -> {
```
```suggestion
            timestampsByLocationByParameter.entrySet().forEach(entry -> {
```

ssalonen replied:
Resolved da5a4367bcdf1904cdc59e3ca76c649dfa62af96

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMIResponse.java:
```diff
@@ -0,0 +1,107 @@
+                timestampsByParameter.entrySet().stream().forEach(entry2 -> {
+                    String parameter = entry2.getKey();
+                    long[] timestamps = entry2.getValue().stream().mapToLong(Long::longValue).toArray();
+                    @Nullable
```
Does this annotation has any effect?

ssalonen replied:
64433ad662b8ac8928f6641f73b8c7cd26bf2ddc

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Request.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    private static String epochToIsoDateTime(long epoch) {
+        return ZonedDateTime.ofInstant(Instant.ofEpochSecond(epoch), ZoneId.of("Z"))
+                .format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"));
```
You could make the DateTimeFormatter a constant to save resources.

ssalonen replied:
Resolved: 4bddbcd8a040b9f43b488ac74b13f10d2a7da298

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+        try {
+            return new Client().queryWeatherStations(STATIONS_TIMEOUT_MILLIS);
+        } catch (FMIResponseException e) {
+            logger.error("Error when querying stations", e);
```
Bindings should only log to error if a bug has been detected, but this catch clause also covers unexpected HTTP error codes. Better log to warn.

ssalonen replied:
6a18547e34848e203d30683f42752617912da7fc

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    @Override
+    @Activate
+    protected void activate(@Nullable Map<String, @Nullable Object> configProperties) {
+        logger.debug("Activating FMI Weather discovery service");
```
Can this message be replaced by using the debugger or use the karaf console logging?

ssalonen replied:
Probably following practice set by other discovery services. Anyways, now the log statement is removed.

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    @Override
+    @Deactivate
+    public void deactivate() {
+        logger.debug("Deactivating FMI Weather discovery service");
```
See above

ssalonen replied:
6a18547e34848e203d30683f42752617912da7fc

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    private void createForecastsForCities(@Nullable PointType currentLocation) {
+        List<Location> filteredCities = new LinkedList<>();
+        CITIES_OF_FINLAND.stream().filter(location2 -> isClose(currentLocation, location2)).peek(station -> {
+            if (logger.isDebugEnabled()) {
```
Are these log messages helpful for any user? Is the result not the same as displayed in PaperUI?

ssalonen replied:
True! fd518b4e9271b71fe63cc07dc0bd1ed98bf007c6

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+        List<Location> candidateStations = new LinkedList<>();
+        List<Location> filteredStations = new LinkedList<>();
+        cachedStations().peek(station -> {
+            if (logger.isDebugEnabled()) {
```
See above

ssalonen replied:
Keeping this one, I'm able to debug users discovery using logs only (spits out candidate stations and filtered/close-by stations)

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+
+    @Override
+    protected void thingDiscovered(DiscoveryResult discoveryResult) {
+        logger.trace("Discovered: {}", discoveryResult);
```
See above

ssalonen replied:
c7cc418704c302cbad2cc527331965501f3ab5f7

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+        super.thingDiscovered(discoveryResult);
+    }
+
+    @SuppressWarnings("null")
```
Better annotate `Set<Location>` in the cache declaration as NonNull.

ssalonen replied:
Unfortunately local variable `stations` can be null still (`ExpiringCache.getValue()` return `@Nullable` value always)

fwolter replied:
If it can be null, wouldn't it be better to apply a null check and remove the supression?

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    protected void stopBackgroundDiscovery() {
+        logger.debug("Stopping FMI Weather background discovery");
+        ScheduledFuture<?> discoveryJob = this.discoveryJob;
+        if (discoveryJob != null && !discoveryJob.isCancelled()) {
```
Canceling a canceled timer is fine.
```suggestion
        if (discoveryJob != null) {
```

ssalonen replied:
b18e6e8bec9faecb35660d854d755b4415668c08

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    @Reference
+    protected void setLocationProvider(LocationProvider locationProvider) {
+        this.locationProvider = locationProvider;
+        logger.debug("LocationProvider set: {}", this.locationProvider);
```
See above

ssalonen replied:
b18e6e8bec9faecb35660d854d755b4415668c08

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+
+    protected void unsetLocationProvider(LocationProvider provider) {
+        this.locationProvider = null;
+        logger.debug("LocationProvider unset");
```
See above

ssalonen replied:
b18e6e8bec9faecb35660d854d755b4415668c08

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    protected LocationProvider getLocationProvider() {
+        logger.trace("getLocationProvider: {}", locationProvider);
```
See above

ssalonen replied:
b18e6e8bec9faecb35660d854d755b4415668c08

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,527 @@
+				<label>Forecast for the current time</label>
+				<description>This is the weather forecast for the current time</description>
+			</channel-group>
+			<channel-group id="forecastHours01" typeId="group-forecast">
```
Yes...it actually came up from user: https://github.com/openhab/openhab-addons/pull/6329#issuecomment-619094384

I don't see any problem having 1h granulatiy.

bjoernbrings replied:
I could think of usecases where you want to include it in some graph

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,527 @@
+		</channel-groups>
+
+		<config-description>
+			<parameter name="fmisid" type="text" required="true">
```
FMI states that FMISID (source, in Finnish: https://www.ilmatieteenlaitos.fi/saahavainnot) is running number, starting from 100000.

I could make it a number indeed with a minimum value -- I guess it would make sense since we get more user interface validation.

ssalonen replied:
Resolved in 72e6e6bf0f68700c0a3c156e4dbd81c581c5995f

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private static int getHours(ChannelUID uid) {
```
I don't understand the question. No -- it's not necessary. But I consider it good practice since it is not using instance data.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,527 @@
+			<channel id="weather-id" typeId="weather-id-channel" />
+		</channels>
+	</channel-group-type>
+	<channel-group-type id="group-forecast-advanced" advanced="true">
```
Discussed earlier: https://github.com/openhab/openhab-addons/pull/6329#issuecomment-618483730

It's just for UI purposes to hide these channels by default in the user interface.

ssalonen commented on bundles/org.openhab.binding.fmiweather/README.md:
```diff
@@ -0,0 +1,1460 @@
+| `precipitation`   | `Number:Length`        | Precipitation in one hour                                                                                                                                                         |
+| `snow-depth`      | `Number:Length`        | Snow depth                                                                                                                                                                        |
+| `visibility`      | `Number:Length`        | Visibility                                                                                                                                                                        |
+| `clouds`          | `Number`               | Cloudiness. Given in numbers out of eight: 0 = clear skies, 8 = overcast, 9 = cloud coverage could not be determined.                                                             |
```
Good point indeed, I guess it makes sense!

ssalonen replied:
Implemented in: 51a22305239e05fe4d3b03f05a5cd1d1df9f7981

Tested manually using couple of stations.

ssalonen commented on bundles/org.openhab.binding.fmiweather/README.md:
```diff
@@ -0,0 +1,1460 @@
+There are two supported things:
+
+- `observation` thing shows current weather observation for a given station. Data is updated automatically every 10 minutes.
+- `forecast` thing shows current weather observation for a given station. Data is updated automatically every 20 minutes.
```
Nicely spotted!

ssalonen replied:
Resolved: c8c2c588a79ec74730766c3a66a12f0f903204c7

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+        addMapping(CHANNEL_FORECAST_WEATHER_ID, PARAM_WEATHER_SYMBOL, null);
+    }
+
+    private @NonNullByDefault({}) LatLon location;
```
I have understood that the way it is know is recommended by the [conventions](https://www.openhab.org/v2.4/docs/developer/development/conventions.html): 

> Fields within ThingHandler classes that are initialized within the initialize() method may also be annotated like this, because the framework ensures that initialize() will be called before any other method.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,193 @@
+        addMapping(CHANNEL_OBSERVATION_PRESENT_WEATHER, PARAM_PRESENT_WEATHER, null);
+    }
+
+    private @NonNullByDefault({}) String fmisid;
```
I have understood that the way it is know is recommended by the [conventions](https://www.openhab.org/v2.4/docs/developer/development/conventions.html): 

> Fields within ThingHandler classes that are initialized within the initialize() method may also be annotated like this, because the framework ensures that initialize() will be called before any other method.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,527 @@
+
+	<!-- Things -->
+	<thing-type id="observation">
+		<label>FMI current weather (observation)</label>
```
Resolved: 7e2d7fd77944d362dd055e5e1abededf5111d67b

ssalonen commented on bundles/org.openhab.binding.fmiweather/README.md:
```diff
@@ -0,0 +1,1460 @@
+Since forecasts are updated at certain times of the day, the last forecast values might be unavailable (`UNDEF`). Typically forecasts between now and 44 hours should be available at all times.
+
+| Channel ID                | Item Type              | Description                                                                                                                                                                                       |     |     |     |
+| ------------------------- | ---------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --- | --- | --- |
```
Resolved: c8c2c588a79ec74730766c3a66a12f0f903204c7

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+            String latlon = location.toString();
+            String[] split = latlon.split(",");
+            this.location = null;
+            this.location = new LatLon(new BigDecimal(split[0].trim()), new BigDecimal(split[1].trim()));
```
It was catched by catch clause. Now having even more user friendly error message in b171a175100dc1d6d1972cbebe24028a06b2bef6

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+            }
+            String latlon = location.toString();
+            String[] split = latlon.split(",");
+            this.location = null;
```
Now location is reseted in dispose:  b171a175100dc1d6d1972cbebe24028a06b2bef6

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,193 @@
+
+    @Override
+    public void initialize() {
+        fmisid = getConfig().get(BindingConstants.FMISID).toString();
```
Resolved: 480414dbb6395c668a09a34775c734af08af2afb

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+                    updateEpochSecondStateIfLinked(channelUID, data.timestampsEpochSecs[timeIndex]);
+                } else {
+                    String field = getDataField(channelUID);
+                    @Nullable
```
Resolved: 480414dbb6395c668a09a34775c734af08af2afb

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,220 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        String.format("%s: %s", e.getClass().getSimpleName(), e.getMessage()));
+                try {
+                    Thread.sleep(RETRY_DELAY_MILLIS);
```
I'm not sure how that would work out really since this method is called with `scheduleWithFixedDelay`.

Assuming exceptions are rare, I find this current approach acceptable.

fwolter replied:
If I see correctly, the scheduler's thread pool is shared amongst all ThingHandlers in the JVM.

I would propose to remove the for-loop and in case of an exception, schedule a one-shot `update()` again with the retry timeout. The retry counter increments, if a one-shot timer is scheduled. I think is can be easily done.

ssalonen replied:
Thanks for checking, then it certainly makes sense to optimize.

ssalonen replied:
Implemented in e455d438856140cd7f0d3435f5165e59af3ddec3 , appreciate second pair of eyes

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,193 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        String.format("%s: %s", e.getClass().getName(), e.getMessage()));
+                try {
+                    Thread.sleep(RETRY_DELAY_MILLIS);
```
I'm not sure how that would work out really since this method is called with `scheduleWithFixedDelay`.

Assuming exceptions are rare, I find this current approach acceptable.

ssalonen replied:
Implemented in e455d438856140cd7f0d3435f5165e59af3ddec3 , appreciate second pair of eyes

ssalonen commented:
Thank you @fwolter for the review, appreciate the time and effort for checking out the code. I've addressed all your comments, please have a another go when you have the time.

fwolter commented on bundles/org.openhab.binding.fmiweather/README.md:
```diff
@@ -0,0 +1,1460 @@
+Number:Length HelsinkiPrecipitation "Precipitation [%.1f %unit%]" <rain> { channel="fmiweather:observation:station_Helsinki_Kumpula:current#precipitation" }
+Number:Length HelsinkiSnowDepth "Snow depth [%.1f %unit%]" <snow> { channel="fmiweather:observation:station_Helsinki_Kumpula:current#snow-depth" }
+Number:Length HelsinkiVisibility "Visibility [%.1f %unit%]" { channel="fmiweather:observation:station_Helsinki_Kumpula:current#visibility" }
+Number HelsinkiClouds "Cloudiness [%.1f %unit%]" { channel="fmiweather:observation:station_Helsinki_Kumpula:current#clouds" }
```
Does 10th percent makes sense here?

fwolter commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,244 @@
+        try {
+            return new Client().queryWeatherStations(STATIONS_TIMEOUT_MILLIS);
+        } catch (FMIResponseException e) {
+            logger.warn("Error when querying stations", e);
```
Would logging the exception's message (instead of the stack trace) sufficient here?

ssalonen replied:
Does that log `cause` of the exception? I think that will be important.

fwolter replied:
Logging the stack trace when the HTTP request times out seems a bit unreasonable. If the cause is important here, why not logging it together with the message?

ssalonen replied:
yes, I was thinking more on the SAX exceptions etc. but good point that timeouts would be logged quite verbosely now.

I'll see what I can do to make it nicer.

ssalonen replied:
All right, now logging only unexpected-content errors (e.g xml errors) verbosely : 6c65be6bb1

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,232 @@
+    private static final int QUERY_RESOLUTION_MINUTES = 20; // The channel group hours should be divisible by this
+    // Hirlam horizon is 54h https://ilmatieteenlaitos.fi/avoin-data-avattavat-aineistot (in Finnish)
+    private static final int FORECAST_HORIZON_HOURS = 50; // should be divisible by QUERY_RESOLUTION_MINUTES
+    private static final Map<String, Map.Entry<String, @Nullable Unit<?>>> CHANNEL_TO_FORECAST_FIELD_NAME_AND_UNIT = new HashMap<>(
```
I see what you are saying. However, I do not want to put `null` entries in this map so I think it is wrong to declare it as `Nullable`. I actually want compiler to disallow entering null-entries to the map.

In my opinion we should not declare the types as null to "workaround" `Map.get`'s wrong definition. Eclipse is saying the following

```
@NonNull Entry<@NonNull String, @Nullable Unit<?>> @NonNull java.util.Map.get(Object key)
```

As we know, it should be always `@Nullable`  not matter how the values are defined.

This is further discussed in https://community.openhab.org/t/null-annotations-warnings-driving-me-nuts-how-to-get-rid-of-them/95187/3

Apart from suppressing warnings in these small methods, I haven't found any other method to keep the right null annotations at declaration level, while getting rid of warnings.

fwolter replied:
I understand your point. It's ok for me.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ObservationWeatherHandler.java:
```diff
@@ -0,0 +1,227 @@
+    public static final Unit<Length> MILLIMETRE = MetricPrefix.MILLI(METRE);
+    public static final Unit<Length> CENTIMETRE = MetricPrefix.CENTI(METRE);
+
+    private static final Map<String, Map.Entry<String, @Nullable Unit<?>>> CHANNEL_TO_OBSERVATION_FIELD_NAME_AND_UNIT = new HashMap<>(
```
Commented above

ssalonen commented:
Thank you @fwolter, nice finds again.

After the reviews it would be perhaps good idea to have some user testing as well, just to be sure no regressions were caused.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,230 @@
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder
+                    .create(new ThingUID(THING_TYPE_OBSERVATION,
+                            cleanId(String.format("station_%s_%s", station.id, station.name))))
+                    .withLabel(String.format("FMI weather observation for %s", station.name)).withProperties(properties)
```
Resolved in c1a9043 if I understood the concept of representation property correctly.

ssalonen commented:
Tested locally and seems to work fine :)

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMISID.java:
```diff
@@ -0,0 +1,41 @@
+@NonNullByDefault
+public class FMISID implements QueryParameter {
+
+    private String fmisid;
```
It might make sense to make it final

ssalonen replied:
9a70fe21c8869e27d56afebd853ed06c28a5766b

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/LatLon.java:
```diff
@@ -0,0 +1,42 @@
+@NonNullByDefault
+public class LatLon implements QueryParameter {
+
+    private String latlon;
```
See other class, consider to make all those simple classes there read-only fields `final`

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Location.java:
```diff
@@ -0,0 +1,74 @@
+     * @param longitude longitude of the location
+     */
+    public Location(String name, String id, BigDecimal latitude, BigDecimal longitude) {
+        super();
```
Can you leave this out

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Location.java:
```diff
@@ -0,0 +1,74 @@
+@NonNullByDefault
+public class Location {
+
+    public final String name;
```
👍

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/CitiesOfFinland.java:
```diff
@@ -0,0 +1,263 @@
+                new Location("Ylöjärvi", "61.55632,23.59606", new BigDecimal("61.55632"), new BigDecimal("23.59606")));
+        CITIES_OF_FINLAND.add(
+                new Location("Ähtäri", "62.55403,24.06186", new BigDecimal("62.55403"), new BigDecimal("24.06186")));
+        CITIES_OF_FINLAND.add(
```
That method is shorter then I expected ;-)

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,226 @@
+        try {
+            return new Client().queryWeatherStations(STATIONS_TIMEOUT_MILLIS);
+        } catch (FMIResponseException e) {
+            if (e instanceof FMIUnexpectedResponseException) {
```
Please turn this in a separate catch block

ssalonen replied:
6f05155ca8dcd4d01be7de0a831e2911e33a4add

martinvw commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java:
```diff
@@ -0,0 +1,226 @@
+    private static final int STATIONS_TIMEOUT_MILLIS = (int) TimeUnit.SECONDS.toMillis(10);
+    private static final int DISCOVER_TIMEOUT_SECONDS = 5;
+    private static final int LOCATION_CHANGED_CHECK_INTERVAL_SECONDS = 60;
+    private static final int FIND_STATION_METERS = 80_000;
```
👍

martinvw commented on bundles/org.openhab.binding.fmiweather/src/test/java/org/openhab/binding/fmiweather/AbstractFMIResponseParsingTest.java:
```diff
@@ -0,0 +1,152 @@
+            return Paths.get(getClass().getResource(filename).toURI());
+        } catch (URISyntaxException e) {
+            fail(e.getMessage());
+            // Make the compiler happy by throwing here, fails already above
```
👍

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+            ScheduledFuture<?> prevFuture = updateChannelsFutureRef.get();
+            ScheduledFuture<?> newFuture = updateChannelsFutureRef
+                    .updateAndGet(fut -> fut == null || fut.isDone() ? submitUpdateChannelsThrottled() : fut);
+            assert newFuture != null; // invariant
```
This is invariant (did you notice the comment?) -- that is, always bound to pass. It is there to aid the developer when reading the code, easier to reason about.

See https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html

If you find this confusing, I can remove it alltogether.

martinvw replied:
I never said I didn't know what it was 😜

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+                        delayRemainingMillis);
+            }
+            if (prevFuture == newFuture) {
+                logger.trace("REFRESH received. Previous refresh ongoing, will wait for it to complete in {} ms",
```
7cba9c40cc55c271f61262bcf0864abaafb17385

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+                logger.trace("REFRESH received. Delaying by {} ms to avoid throttle excessive REFRESH",
+                        delayRemainingMillis);
+            }
+            if (prevFuture == newFuture) {
```
7cba9c40cc55c271f61262bcf0864abaafb17385

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+        long nextRefresh = lastRefreshMillis + REFRESH_THROTTLE_MILLIS;
+        lastRefreshMillis = now;
+        if (now > nextRefresh) {
+            return scheduler.schedule(this::updateChannels, 0, TimeUnit.MILLISECONDS);
```
I do need ScheduledFuture interface for `getDelay` method.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+        } else {
+            // logger.error(messageIfNotPresent, args) avoided due to static analyzer
+            String formattedMessage = String.format(messageIfNotPresent, args);
+            logger.error("Unwrapping error: {}", formattedMessage);
```
OFFLINE/ONLINE status was already handled by the catch clause, but I see your point.

Changed unwrap to use FMIUnexpectedResponseException since all failing unwraps are due to unexpected content in the API response. Using warn level.

See a17d51c68be3a756d4de83bb2c1793c5768b9ba6

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java:
```diff
@@ -0,0 +1,233 @@
+            // logger.error(messageIfNotPresent, args) avoided due to static analyzer
+            String formattedMessage = String.format(messageIfNotPresent, args);
+            logger.error("Unwrapping error: {}", formattedMessage);
+            throw new IllegalStateException("unwrapping");
```
See comment above, and commit a17d51c68be3a756d4de83bb2c1793c5768b9ba6

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+        try {
+            Object location = getConfig().get(BindingConstants.LOCATION);
+            if (location == null) {
+                throw new IllegalStateException();
```
Improved in cba19a5dd154923dbd2fdd4895e4a3fcceac96eb

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+        } catch (IllegalStateException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    String.format("location parameter not set %s %s", e.getClass().getName(), e.getMessage()));
+        } catch (NumberFormatException | IndexOutOfBoundsException e) {
```
See 35b05257464e40dc971164f2b8ede609af179154

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+            String latlon = location.toString();
+            String[] split = latlon.split(",");
+            if (split.length != 2) {
+                throw new NumberFormatException(String.format(
```
We still have to catch NumberFormatException -- so I don't see how the comment on performance applies?

martinvw replied:
Catching is not the performance impact, it's the throwing side of things, but assuming this an exceptional path it is not blocking for me. But the general rule is that if you handle an exception in the same method as were its throw your method is either too big or it should have been a normal if-condition or sometimes both 🙂

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+                response = null;
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        String.format("%s: %s", e.getClass().getSimpleName(), e.getMessage()));
+                logger.trace("Query failed. Increase retry count {} and try again. Error: {} {}", retry,
```
I think previous reviewer did not like stack traces, not even trace level. I think everything that is needed will be logged now.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+                                QUERY_RESOLUTION_MINUTES),
+                        QUERY_RESOLUTION_MINUTES), TIMEOUT_MILLIS);
+            } catch (FMIResponseException e) {
+                if (e instanceof FMIUnexpectedResponseException) {
```
427d91bce02f1cd873c05cbd26898225a4587fbb

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+        if (retry < RETRIES) {
+            try {
+                long now = Instant.now().getEpochSecond();
+                response = client.query(new ForecastRequest(location, floorToEvenMinutes(now, QUERY_RESOLUTION_MINUTES),
```
updateChannels is also called with REFRESH command -- we use cached response in that case. In case response is updated in the meanwhile (if this is what you referred with unsafge), I don't see it a problem really.

martinvw replied:
Given it is also called from another path it is a false-positive 😄

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+                    String field = getDataField(channelUID);
+                    Unit<?> unit = getUnit(channelUID);
+                    if (field == null) {
+                        logger.error("Channel {} not handled. Bug?", channelUID.getId());
```
Is this an issue really? This can only occur with broken code -- when channelUID is not in static constant  CHANNEL_TO_FORECAST_FIELD_NAME_AND_UNIT

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/ForecastWeatherHandler.java:
```diff
@@ -0,0 +1,238 @@
+        return (int) (TimeUnit.HOURS.toMinutes(hours) / QUERY_RESOLUTION_MINUTES);
+    }
+
+    @SuppressWarnings({ "unused", "null" })
```
Suppression refers to "dead code" "return null" part of 
```
        if (entry == null) {
            return null;
        }
```
This again is false alarm (that is, it is not dead code!), root cause being buggy nullness definition of Map.get. See above discussion on this as well.

martinvw replied:
(Y) I know that pain only forgot/did not realize it will give also give "unused", I'm an IntelliJ user so I do not see the warnings in-line.

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/Client.java:
```diff
@@ -0,0 +1,445 @@
+        });
+
+        return builder.build();
+    }
```
Unfortunately I do not see it easily splitted

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMIResponse.java:
```diff
@@ -0,0 +1,105 @@
+        public FMIResponse build() {
+            Map<Location, Map<String, Data>> out = new HashMap<>(timestampsByLocationByParameter.size());
+
+            timestampsByLocationByParameter.entrySet().forEach(entry -> {
```
6a56f3a61378b595b7f306ed1af3fafe7a086f4d

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMIResponse.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    private Map<Location, Map<String, Data>> dataByLocationByParameter;
```
bb09186013c816145d44cfbdff77a726729f5bd5

ssalonen commented on bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/client/FMIResponse.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public String toString() {
+        return new StringBuilder("FMIResponse(").append(dataByLocationByParameter.toString()).append(")").toString();
```
What would be your preference instead?

martinvw replied:
Generating them! So given you did that, it's not a problem. Otherwise, I would just use string concatenation that is only bad in loops.

ssalonen commented:
Thank you @martinvw a plenty for the review. I've addressed your comments above. Please have another go when you have the time.

martinvw commented:
It would be nice if you could align the logging statements with the documented rules regarding the error level. I suppose it would be best to do that in a separate PR. If you want to do it now it let me now quickly :-)

martinvw commented:
@ssalonen thanks for all your hard work.
